### PR TITLE
A complete rewrite.

### LINKED ISSUE #18: Rewrite: Complete rewrite of the tool.
I've decided that this tool sucks, the code is bad, it's buggy on Windows and has problems here and there. So, I'm rewriting it, this time with a bit more overhead and hopefully more skill.

- [x] Entry View Window.
- [x] Standard upload functionality.
- [x] Edit existing entries
- [x] Advanced options.
- [x] Comprehensive Error Display.
- [x] Cleanup and Refactor. 

If there should be more added to this issue, or like to discuss the points listed, please reply to this Issue.

The rewrite will happen on the `new/rewrite` branch.

### PR DESCRIPTION
**Pull request by Trico-Everfire**
The old code was bad, it was riddled with bugs and I am surprised it even "worked" under some circumstances. But! I have improved it! massively. and this should fix most of the issues.

will resolve #18

### CODE REVIEW

File: src/mapuploader.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		void addTagWidgetItem( const QString &name, bool removable = true );
+		void populateDefaultTagListWidget() const;
+		void simpleInputDialog( QString &resultString );
+		QListWidgetItem *createBasicListWidgetItem( const QString &labelText, QWidget *baseWidget, QListWidget *itemList, ListItemTypes type, bool removable = true );
+	};
+
+	class CMapUploader : public QDialog
+	{
+	public:
+		static constexpr int MAX_IMAGE_SIZE = 1048576;
```
Comment (craftablescience) [inline src/mapuploader.h:57]:
```suggestion
		static constexpr int MAX_IMAGE_SIZE = 1024 * 1024; // 1mb
```

File: src/mapuploader.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+		// We need to keep the strings alive long enough for
+		// m_CallOldApiResultSubmitItemUpdate to process them.
+		struct steamTagListPreserver
```
Comment (craftablescience) [inline src/mapuploader.h:61]: This is the camelCase class I was talking about earlier

**Review by craftablescience — COMMENTED** [review]
Looked it over, vast majority is good code, you've come a long way

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 name: Build
 on:
   push:
-    branches: [ master ]
+    branches: [ master, fix/thumbnail-fix, new/rewrite ]
```
[the remark this answers is not in the document]
Comment (Trico-Everfire) [reply .github/workflows/build.yml:4]: Cannot accept due to PRs unable to use secret keys (stupid gh)

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,101 +1,105 @@
...
-	# Copy these to bundle them with the program in releases
-	configure_file("${QT_BASEDIR}/bin/Qt5Core.dll"    "${CMAKE_BINARY_DIR}/Qt5Core.dll"    COPYONLY)
-	configure_file("${QT_BASEDIR}/bin/Qt5Gui.dll"     "${CMAKE_BINARY_DIR}/Qt5Gui.dll"     COPYONLY)
-	configure_file("${QT_BASEDIR}/bin/Qt5Widgets.dll" "${CMAKE_BINARY_DIR}/Qt5Widgets.dll" COPYONLY)
-	configure_file("${QT_BASEDIR}/bin/Qt5Network.dll" "${CMAKE_BINARY_DIR}/Qt5Network.dll" COPYONLY)
-	configure_file("${QT_BASEDIR}/plugins/platforms/qwindows.dll"         "${CMAKE_BINARY_DIR}/platforms/qwindows.dll"         COPYONLY)
-	configure_file("${QT_BASEDIR}/plugins/platforms/qwindowsd.dll"        "${CMAKE_BINARY_DIR}/platforms/qwindowsd.dll"        COPYONLY)
-endif()
\ No newline at end of file
+option(USE_STEAM "Uses steam" OFF)
```
[the remark this answers is not in the document]
Comment (Trico-Everfire) [reply CMakeLists.txt:42]: Leftover from Erin's base cmakelist.txt, will be removed.

File: src/mapuploader.cpp [hunk 5]
```diff
@@ -0,0 +1,986 @@
...
+		auto text = QString( item->data( Qt::UserRole ).toString() );
+
+		std::shared_ptr<char[]> charUPtr( new char[text.length()] );
+		strcpy( charUPtr.get(), text.toLocal8Bit().constData() );
+		m_pSteamTagList.m_SharedCharArrayPointerTagList.push_back( charUPtr );
+		charTagVector.push_back( charUPtr.get() );
+	}
+
+	m_pSteamTagList.m_SteamTagList.m_nNumStrings = (int32)charTagVector.size();
+	m_pSteamTagList.m_SteamTagList.m_ppStrings = (const char **)charTagVector.data();
```
[the remark this answers is not in the document]
Comment (Trico-Everfire) [reply src/mapuploader.cpp:482]: Huh? this works fine, where does the nullptr come in?

[the remark this answers is not in the document]
Comment (craftablescience) [reply src/mapuploader.cpp:482]: Ignore that, didn't see it had a parameter for size

File: src/mapuploader.cpp [hunk 6]
```diff
@@ -0,0 +1,986 @@
...
+{
+	auto updateItemHandle = SteamUGC()->StartItemUpdate( CMainView::m_GameID, publishedFileId );
+
+	SteamUGC()->SetItemTitle( updateItemHandle, m_pTitleLine->text().toLocal8Bit().constData() );
+	if ( !m_pDescLine->toPlainText().isEmpty() )
+		SteamUGC()->SetItemDescription( updateItemHandle, m_pDescLine->toPlainText().toLocal8Bit().constData() );
+
+	SteamUGC()->SetItemVisibility( updateItemHandle, static_cast<ERemoteStoragePublishedFileVisibility>( m_pAdvancedOptions->m_pVisibilityComboBox->currentData( Qt::UserRole ).toInt() ) );
+
+	auto charTagVector = std::vector<const char *> {};
```
[the remark this answers is not in the document]
Comment (craftablescience) [reply src/mapuploader.cpp:466]: I guess as long as it works it's not too much extra code

**Pull request merged**